README.md

```diff
@@ -219,54 +367,171 @@ for (padded_string json : { string("[1, 2, 3]"), string("true"), string("[ true,
+cout << "simdjson v" << #SIMDJSON_VERSION << endl;
+cout << "Detected the best implementation for your machine: " << simdjson::active_implementation->name();
+cout << "(" << simdjson::active_implementation->description() << ")" << endl;
+```
 
-for (int i=0;i<argc;i++) {
-  auto [doc, error] = parser.parse(get_corpus(argv[i]));
-  if (error == CAPACITY) { cerr << "JSON files larger than 1MB are not supported!" << endl; exit(1); }
-  if (error) { cerr << error << endl; exit(1); }
-  cout << doc;
+Implementation detection will happen in this case when you first call `name()`.
+
+#### Querying Available Implementations
+
+You can list all available implementations:
+
+```c++
+for (auto implementation : simdjson::available_implementations) {
+  cout << implementation->name() << ": " << implementation->description() << endl;
 }
 ```
 
-## Newline-Delimited JSON (ndjson) and  JSON lines
+And look them up by name:
 
-The simdjson library also support multithreaded JSON streaming through a large file containing many smaller JSON documents in either [ndjson](http://ndjson.org) or [JSON lines](http://jsonlines.org) format. If your JSON documents all contain arrays or objects, we even support direct file concatenation without whitespace. The concatenated file has no size restrictions (including larger than 4GB), though each individual document must be less than 4GB.
+```c++
+cout << simdjson::available_implementations["fallback"]->description() << endl;
+```
 
-Here is a simple example, using single header simdjson:
+#### Manually Selecting the Parser Implementation
 
-```cpp
-#include "simdjson.h"
-#include "simdjson.cpp"
+If you're trying to do performance tests or see how different implementations of simdjson run, you
+can select the CPU architecture yourself:
 
-int parse_file(const char *filename) {
-    simdjson::document::parser parser;
-    for (const document &doc : parser.load_many(filename)) {
-      // do something with the document ...
-    }
-}
+```c++
+// Use the fallback implementation, even though my machine is fast enough for anything
+simdjson::active_implementation = simdjson::available_implementations["fallback"];
 ```
 
-## Usage: easy single-header version
+### Large files and huge page support
+
+If you are processing large files (e.g., 100 MB), it is possible that the performance of simdjson will be limited by page misses and/or page allocation. [On some systems, memory allocation runs far slower than we can parse (e.g., 1.4GB/s).](https://lemire.me/blog/2020/01/14/how-fast-can-you-allocate-a-large-block-of-memory-in-c/)
+
+A viable strategy is to amortize the cost of page allocation by reusing the same `parser` object over several files:
+
+```C++
+// create one parser
+simdjson::document::parser parser;
+...
+// the parser is going to pay a memory allocation price
+auto [doc1, error1] = parser.parse(largestring1);
+...
+// use again the same parser, it will be faster
+auto [doc2, error2] = parser.parse(largestring2);
```

Comment:
If doc1 and doc2 are both references, are they actually references to the same thing? If not, what is the state of doc1... I think that a reader might want to know.